scapy/contrib/geneve.py

```diff
@@ -20,29 +18,49 @@
 """
 Geneve: Generic Network Virtualization Encapsulation
 
-draft-ietf-nvo3-geneve-06
+draft-ietf-nvo3-geneve-16
 """
 
+import struct
+
 from scapy.fields import BitField, XByteField, XShortEnumField, X3BytesField, \
-    XStrField
+    XStrField, XShortField, StrField, XByteField, FieldLenField, PacketListField
 from scapy.packet import Packet, bind_layers
 from scapy.layers.inet import IP, UDP
 from scapy.layers.inet6 import IPv6
 from scapy.layers.l2 import Ether, ETHER_TYPES
 from scapy.compat import chb, orb
 from scapy.error import warning
 
+CLASS_IDS = {0x0100: "Linux",
+             0x0101: "Open vSwitch",
+             0x0102: "Open Virtual Networking (OVN)",
+             0x0103: "In-band Network Telemetry (INT)",
+             0x0104: "VMware",
+             0x0105: "Amazon.com, Inc.",
+             0x0106: "Cisco Systems, Inc.",
+             0x0107: "Oracle Corporation",
+             0x0110: "Amazon.com, Inc.",
+             0x0118: "IBM",
+             0x0128: "Ericsson",
+             0xFEFF: "Unassigned",
+             0xFFFF: "Experimental"}
 
-class GENEVEOptionsField(XStrField):
-    islist = 1
-
-    def getfield(self, pkt, s):
-        opln = pkt.optionlen * 4
-        if opln < 0:
-            warning("bad optionlen (%i). Assuming optionlen=0", pkt.optionlen)
-            opln = 0
-        return s[opln:], self.m2i(pkt, s[:opln])
+class GeneveOptions(Packet):
+    name = "Geneve Options"
+    fields_desc = [XShortEnumField("options_class", 0x0000, CLASS_IDS),
```

Comment:
I will prefer to use the field names: class, type, reserved, length and data to be consistent with the draft. You should use `Bitfield("reserved", 0, 3)` instead of the three `BitField`.

Reply:
Sure, will modify and re submit